### DISCUSSION TITLE
Experiments with tanh at criticality

### DISCUSSION THREAD

**Discussion by vkmbox** [vkmbox/Ml-readinggroup#5] [Show and tell]
Hi all,

Just uploaded to master a new version of [relu-experiments](https://github.com/vkmbox/Ml-readinggroup/blob/master/Roberts_Yaida/chapter5/init_distribution_ffn_relu.ipynb) and a notebook with [tanh's at criticality](https://github.com/vkmbox/Ml-readinggroup/blob/master/Roberts_Yaida/chapter5/init_distribution_ffn_tanh.ipynb). Some conclusions:
1) In relu-experiments I increased number of experiments to 200 and covariation-values become more predictable. G00 and G11 are approximately stable from input to output as predicted in criticality-mode and respective covariations on FFN output are approximately the same. Between-newron covariation becomes small < 0.1 which is a sign of newron's independence, probably I'll check the independence with other criteria. Only the problem with relu experiments is a growing G01 which must be stable according to the theory I think.
2) In tanh-experiments an exponential decay of the initialisation's variance and G-metric is observed. I set all layer's width $n_l$ including input and output equal to 10000 and initialised weights $C_b$=0 and $C_w$ according to (5.139) and (5.142). However, signal is not stable and decays for all G00, G01 and G11. This contradicts theory, @Bgzh, @Brbauwens, can you pls check if I'm missing something.

**Comment by Brbauwens** [comment]
In ReLu, G01 should be growing when the depth increases. (Or are you talking about growing when the width n increases?) I hope to read a bit more in the book by our meeting on next Saturday, and then plan to check your code.


For reference I place here some old email conversation. 

> Just noticed at the beginning of p. 111 the K_xx is a kernel-part of G_xx. Bruno, can you confirm that matrix (5.15) is like a covariance matrix for a preactivation on l-level for +/- train-points (up to NLO, etc) or it's a wrong assumption?

Yes, it is correct, the Kernel is the leading order part of the matrix G_{alpha1,alpha2}, it is defined in section 4.105 on page 100.

> Also, there is an analysis on p. 140 saying that for relu with finite-width NLO is zero, consequently K_xx is almost G_xx. Regarding 5.166 - does it mean G_01 goes to zero?

On page 140, I only see discussions about the first layer (around equation 5.118). But in general if we may assume that the width is large, it means that to discover trends, we should first look at the leading order terms of G, (which are denoted with K). Note that \phi_\ell goes to zero, and K_{+-} is proportional to K_{00} \cos(\phi_\ell), thus K_{+-} approaches K_{00} for large \ell (this is what I mean in my previous email).

>  Also, 5.166 uses a diagonal kernel which is not the case in this experiment, for input generation a standard normal distribution is used.

Yes, this assumption probably does not apply (I still need to look at the python file). In chapter 5, this means that they set the coordinate K_1 = 0, thus \delta K_1 = 0.

On the other hand, when the changes of K^{(\ell)} after going to the next layer are not too big, one may hope that various contributions in K^{(\ell)} add up independently, and I think we should look at equation (5.48) on page 121, but this formula does not apply to ReLu, so we need something else. But I expect that the 2nd term in (5.48) is still correct, and that the function h(K) is positive.  This may explain why K^{(\ell)}_{+-}/K^{(\ell)}_{00} increases with \ell. But I do not have all this fresh in my memory...

**Reply by vkmbox** [reply]
>we should look at equation (5.48) on page 121, but this formula does not apply to ReLu

For (5.46)-(5.48) at criticality in relu-case there is a decision (5.63)-(5.65) under the condition (5.62) with сhi=1. Thus, left-side in (5.45) should be approx. constant and G01 should be stable as well, I don't understand why G01 is expected to be growing.

**Reply by Brbauwens** [reply]
According to equation 5.99 on page 134, we have that K_[2] is decreasing. If K_++ and K_-- are constant, that means that K_+- is increasing.

**Reply by vkmbox** [reply]
In relu-experiments we work with "Scale-Invariant" class whereas 5.99 is for "K = 0 Universality" class. For Scale-Invariant we have K_[2] from (5.65), given Chi is one we have K_[2] is constant.

**Reply by vkmbox** [reply]
As discussed, in tanh-experiments $G_{xx}$-metrics decay but rather polinomially then exponentially, which is as expected according to (5.94). A notebook with [tanh's at criticality](https://github.com/vkmbox/Ml-readinggroup/blob/master/Roberts_Yaida/chapter5/init_distribution_ffn_tanh.ipynb) is changed correspondently.
